File: libs/observability/src/cost_model.rs
//! Foundry compute-seconds cost model for media-set transformations.
//!
//! Pinned line-by-line against the table in
//! `Data formats/Media sets (unstructured data)/Media usage costs and limits.md`
//! ("Transformations" section). The numbers below are the Foundry-published
//! `compute-seconds per GB processed` rates; the `media-transform-runtime-service`
//! consults this module for every access-pattern invocation, multiplies by the
//! input size in GB, and emits the result into:
//!
//!   * `media_compute_seconds_total{transformation, schema}`
//!     (Prometheus counter declared in `services/media-sets-service/src/metrics.rs`).
//!   * The `media_set.access_pattern_invoked` audit event
//!     ([`audit_trail::events::AuditEvent::MediaSetAccessPatternInvoked`]).
//!
//! ## Why this lives in `libs/observability`
//!
//! Multiple services need to read the same table:
//!   * `media-transform-runtime-service` — the actual worker, charges
//!     compute time per call.
//!   * `media-sets-service` — surfaces the cost in the `GET /usage`
//!     endpoint that powers the Usage UI tab.
//!   * Future planning / quota tooling (`tenancy-organizations-service`,
//!     billing exports).
//!
//! Centralising the table here means a Foundry doc update touches one
//! file and a single snapshot test (`cost_model_matches_table.rs`)
//! enforces drift never sneaks past CI.

use serde::{Deserialize, Serialize};

/// Foundry media-set schema discriminator. Mirrors the upper-snake-case
/// proto enum strings (`IMAGE`, `AUDIO`, `VIDEO`, `DOCUMENT`,
/// `SPREADSHEET`, `EMAIL`) used everywhere else in the platform.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum SchemaKind {
    All,
    Image,
    Audio,
    Video,
    Document,
    Spreadsheet,
}

impl SchemaKind {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::All => "ALL",
            Self::Image => "IMAGE",
            Self::Audio => "AUDIO",
            Self::Video => "VIDEO",
            Self::Document => "DOCUMENT",
            Self::Spreadsheet => "SPREADSHEET",
        }
    }
}

/// Single row of the published cost table.
///
/// `key` is the canonical wire-form transformation name we use across
/// the cost meter, the audit envelope `access_pattern` field, and the
/// `transformation` Prometheus label. `display_name` reproduces the
/// Foundry doc verbatim so the Usage UI can show the same caption an
/// operator would see in the upstream documentation.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct CostEntry {
    /// Wire-form key: lowercase, snake-cased. Stable across releases —
    /// changing it breaks the audit envelope category and the
    /// Prometheus label cardinality.
    pub key: &'static str,
    /// Foundry doc display name (verbatim).
    pub display_name: &'static str,
    pub schema: SchemaKind,
    /// Compute-seconds **per gigabyte** of input processed.
    pub compute_seconds_per_gb: u32,
}

/// Foundry "All" row — applies to download / streaming of bytes
/// regardless of media schema.
pub const DOWNLOAD_STREAM: CostEntry = CostEntry {
    key: "download_stream",
    display_name: "Download / stream",
    schema: SchemaKind::All,
    compute_seconds_per_gb: 2,
};

/// The full table. Order mirrors the doc top-to-bottom so a screen-by-
/// screen review against the upstream rendering catches a missing /
/// reordered row instantly. **DO NOT REORDER without updating the
/// `cost_model_matches_table` snapshot test.**
pub const COST_TABLE: &[CostEntry] = &[
    // ── All ───────────────────────────────────────────────────────
    DOWNLOAD_STREAM,
    // ── Images ────────────────────────────────────────────────────
    CostEntry { key: "generate_pdf",                  display_name: "Generate PDF",                  schema: SchemaKind::Image, compute_seconds_per_gb: 40 },
    // `thumbnail` is documented in `Transforming media.md` as a
    // canonical Foundry use case ("Thumbnails and previews for PDFs
    // in Workshop") but is not separately billed in the published
    // cost table — Foundry charges thumbnails at the resize rate.
    // We surface it explicitly here so the runtime catalog can
    // dispatch it natively without inventing an unbillable kind.
    CostEntry { key: "thumbnail",                     display_name: "Thumbnail (alias of Resize)",   schema: SchemaKind::Image, compute_seconds_per_gb: 40 },
    CostEntry { key: "resize",                        display_name: "Resize",                        schema: SchemaKind::Image, compute_seconds_per_gb: 40 },
    CostEntry { key: "resize_within_bounding_box",    display_name: "Resize within bounding box",    schema: SchemaKind::Image, compute_seconds_per_gb: 40 },
    CostEntry { key: "rotate",                        display_name: "Rotate",                        schema: SchemaKind::Image, compute_seconds_per_gb: 40 },
    CostEntry { key: "adjust_contrast",               display_name: "Adjust contrast",               schema: SchemaKind::Image, compute_seconds_per_gb: 75 },
    CostEntry { key: "annotate",                      display_name: "Annotate",                      schema: SchemaKind::Image, compute_seconds_per_gb: 75 },
    CostEntry { key: "crop",                          display_name: "Crop / chip",                   schema: SchemaKind::Image, compute_seconds_per_gb: 75 },
    CostEntry { key: "encryption_decryption",         display_name: "Encryption / decryption",       schema: SchemaKind::Image, compute_seconds_per_gb: 75 },
    CostEntry { key: "geo_tile",                      display_name: "Geo tile",                      schema: SchemaKind::Image, compute_seconds_per_gb: 75 },
    CostEntry { key: "grayscale",                     display_name: "Grayscale",                     schema: SchemaKind::Image, compute_seconds_per_gb: 75 },
    CostEntry { key: "render_dicom_image_layer",      display_name: "Render DICOM image layer",      schema: SchemaKind::Image, compute_seconds_per_gb: 75 },
    CostEntry { key: "ocr",                           display_name: "Extract text (OCR)",            schema: SchemaKind::Image, compute_seconds_per_gb: 275 },
    CostEntry { key: "embedding",                     display_name: "Generate embedding",            schema: SchemaKind::Image, compute_seconds_per_gb: 275 },
    CostEntry { key: "ocr_v2",                        display_name: "Extract text v2 (OCR)",         schema: SchemaKind::Image, compute_seconds_per_gb: 275 },
    CostEntry { key: "extract_layout_aware_v2",       display_name: "Extract layout aware v2",       schema: SchemaKind::Image, compute_seconds_per_gb: 275 },
    // ── Audio ─────────────────────────────────────────────────────
    CostEntry { key: "chunk",                         display_name: "Chunk",                         schema: SchemaKind::Audio, compute_seconds_per_gb: 75 },
    CostEntry { key: "channel_select",                display_name: "Select channel",                schema: SchemaKind::Audio, compute_seconds_per_gb: 75 },
    CostEntry { key: "hls_stream",                    display_name: "Stream with HLS",               schema: SchemaKind::Audio, compute_seconds_per_gb: 75 },
    CostEntry { key: "transcode",                     display_name: "Transcode",                     schema: SchemaKind::Audio, compute_seconds_per_gb: 75 },
    CostEntry { key: "waveform",                      display_name: "Waveform generation",           schema: SchemaKind::Audio, compute_seconds_per_gb: 75 },
    CostEntry { key: "transcription",                 display_name: "Transcription",                 schema: SchemaKind::Audio, compute_seconds_per_gb: 275 },
    // ── Video ─────────────────────────────────────────────────────
    CostEntry { key: "scene_frames_timestamps",       display_name: "Get timestamps for scene frames", schema: SchemaKind::Video, compute_seconds_per_gb: 40 },
    CostEntry { key: "extract_audio",                 display_name: "Extract audio",                 schema: SchemaKind::Video, compute_seconds_per_gb: 75 },
    CostEntry { key: "extract_frames",                display_name: "Extract frames at timestamp",   schema: SchemaKind::Video, compute_seconds_per_gb: 75 },
    CostEntry { key: "video_chunk",                   display_name: "Chunk",                         schema: SchemaKind::Video, compute_seconds_per_gb: 275 },
    CostEntry { key: "scene_frames_all",              display_name: "Extract all scene frames",      schema: SchemaKind::Video, compute_seconds_per_gb: 275 },
    CostEntry { key: "video_hls_stream",              display_name: "Stream with HLS",               schema: SchemaKind::Video, compute_seconds_per_gb: 275 },
    CostEntry { key: "video_transcode",               display_name: "Transcode",                     schema: SchemaKind::Video, compute_seconds_per_gb: 275 },
    // ── Documents ─────────────────────────────────────────────────
    CostEntry { key: "pdf_page_dimensions",           display_name: "Get PDF page dimensions",       schema: SchemaKind::Document, compute_seconds_per_gb: 40 },
    CostEntry { key: "render_page",                   display_name: "Render page as image",          schema: SchemaKind::Document, compute_seconds_per_gb: 40 },
    CostEntry { key: "render_page_bbox",              display_name: "Render page as image within bounding box", schema: SchemaKind::Document, compute_seconds_per_gb: 40 },
    CostEntry { key: "extract_text_raw",              display_name: "Extract all text (raw)",        schema: SchemaKind::Document, compute_seconds_per_gb: 75 },
    CostEntry { key: "extract_form_fields",           display_name: "Extract form fields",           schema: SchemaKind::Document, compute_seconds_per_gb: 75 },
    CostEntry { key: "extract_toc",                   display_name: "Extract table of contents",     schema: SchemaKind::Document, compute_seconds_per_gb: 75 },
    CostEntry { key: "extract_text_pages_raw",        display_name: "Extract text from pages to array (raw)", schema: SchemaKind::Document, compute_seconds_per_gb: 75 },
    CostEntry { key: "extract_text_on_page_raw",      display_name: "Extract text on page (raw)",    schema: SchemaKind::Document, compute_seconds_per_gb: 75 },
    CostEntry { key: "slice_pdf_range",               display_name: "Slice PDF range",               schema: SchemaKind::Document, compute_seconds_per_gb: 75 },
    CostEntry { key: "doc_ocr",                       display_name: "Extract text (OCR)",            schema: SchemaKind::Document, compute_seconds_per_gb: 275 },
    CostEntry { key: "extract_text_pages_ocr",        display_name: "Extract text from pages to array (OCR)", schema: SchemaKind::Document, compute_seconds_per_gb: 275 },
    CostEntry { key: "layout_aware_v2",               display_name: "Extract layout aware text (raw / OCR)", schema: SchemaKind::Document, compute_seconds_per_gb: 275 },
    CostEntry { key: "vlm_extract",                   display_name: "Extract text using VLM",        schema: SchemaKind::Document, compute_seconds_per_gb: 275 },
    // ── Spreadsheets ──────────────────────────────────────────────
    CostEntry { key: "render_sheet",                  display_name: "Convert sheet to JSON",         schema: SchemaKind::Spreadsheet, compute_seconds_per_gb: 275 },
];

/// Lookup the rate for a transformation key. Returns `None` if the key
/// is not in the published table — callers should refuse to charge for
/// an off-table transformation rather than guess. The runtime catalog
/// is the single source of truth for which keys are billable.
pub fn rate_per_gb(key: &str) -> Option<u32> {
    COST_TABLE
        .iter()
        .find(|entry| entry.key == key)
        .map(|entry| entry.compute_seconds_per_gb)
}

/// Compute-seconds charged for processing `bytes` through the
/// transformation identified by `key`. Rounds **up** to the nearest
/// whole compute-second so a 1-byte invocation still charges at least
/// the entry's GB rate × byte fraction (rounded up). Returns `None` if
/// `key` is not in the table.
pub fn charge_compute_seconds(key: &str, bytes: u64) -> Option<u64> {
    let rate = rate_per_gb(key)? as u64;
    // Compute-seconds = rate × (bytes / GB). Use 64-bit math throughout
    // to keep multi-GB inputs precise.
    const GB: u64 = 1024 * 1024 * 1024;
    let charged = (rate.saturating_mul(bytes) + GB - 1) / GB;
    // Tiny inputs still incur measurable cost (the table publishes the
    // GB rate; we expose at minimum 1 compute-second for any non-zero
    // request so the meter never reports a 0-charge invocation).
    Some(charged.max(if bytes == 0 { 0 } else { 1 }))
}

/// Convenience: lookup the entry for a key, including its display name
/// and schema. Used by the `/usage` REST surface and the Usage UI tab.
pub fn entry(key: &str) -> Option<CostEntry> {
    COST_TABLE.iter().copied().find(|entry| entry.key == key)
}

/// Iterator over every transformation belonging to a given schema
/// (plus the `All` rows). Backs the `/usage` JSON breakdown so the UI
/// can populate the per-kind chart without knowing the table layout.
pub fn entries_for(schema: SchemaKind) -> impl Iterator<Item = &'static CostEntry> {
    COST_TABLE
        .iter()
        .filter(move |entry| entry.schema == schema || entry.schema == SchemaKind::All)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn download_stream_rate_matches_doc() {
        assert_eq!(rate_per_gb("download_stream"), Some(2));
    }

    #[test]
    fn ocr_charges_275_per_gb_in_image_schema() {
        assert_eq!(rate_per_gb("ocr"), Some(275));
        assert_eq!(entry("ocr").unwrap().schema, SchemaKind::Image);
    }

    #[test]
    fn unknown_key_is_not_billable() {
        assert_eq!(rate_per_gb("not_a_real_transformation"), None);
        assert_eq!(charge_compute_seconds("not_a_real_transformation", 1024), None);
    }

    #[test]
    fn charge_rounds_up_and_respects_minimum() {
        // 1 GB through `download_stream` = 2 compute-seconds exactly.
        assert_eq!(
            charge_compute_seconds("download_stream", 1024 * 1024 * 1024),
            Some(2)
        );
        // 1 byte must still register a non-zero charge (rounded to 1).
        assert_eq!(charge_compute_seconds("download_stream", 1), Some(1));
        // 0 bytes processes nothing.
        assert_eq!(charge_compute_seconds("download_stream", 0), Some(0));
    }

    #[test]
    fn entries_for_image_includes_all_schema_rows() {
        // `All` rows (download_stream) must appear under every schema
        // so the per-kind UI breakdown is self-contained.
        let keys: Vec<&str> = entries_for(SchemaKind::Image)
            .map(|e| e.key)
            .collect();
        assert!(keys.contains(&"download_stream"), "All rows must be visible");
        assert!(keys.contains(&"ocr"), "Image rows must be visible");
        assert!(!keys.contains(&"transcription"), "Audio rows must NOT leak");
    }
}

File: libs/observability/costmodel/costmodel.go
// Package costmodel mirrors the Rust libs/observability/src/cost_model.rs
// Foundry compute-seconds cost table for media-set transformations.
//
// Pinned line-by-line against the table in
// `Data formats/Media sets (unstructured data)/Media usage costs and limits.md`
// ("Transformations" section). The numbers below are the Foundry-published
// `compute-seconds per GB processed` rates; media-transform-runtime-service
// consults this package for every access-pattern invocation, multiplies by
// the input size in GB, and emits the result into:
//
//   - media_compute_seconds_total{transformation, schema} (Prometheus
//     counter declared in media-sets-service).
//   - The media_set.access_pattern_invoked audit event.
//
// Centralising the table here means a Foundry doc update touches one
// file and a single snapshot test enforces drift never sneaks past CI.
package costmodel

// SchemaKind discriminates the Foundry media-set schema. Wire form
// matches the proto enum (SCREAMING_SNAKE_CASE) used everywhere else
// in the platform.
type SchemaKind string

const (
	SchemaAll         SchemaKind = "ALL"
	SchemaImage       SchemaKind = "IMAGE"
	SchemaAudio       SchemaKind = "AUDIO"
	SchemaVideo       SchemaKind = "VIDEO"
	SchemaDocument    SchemaKind = "DOCUMENT"
	SchemaSpreadsheet SchemaKind = "SPREADSHEET"
)

// CostEntry is a single row of the published cost table.
//
// Key is the canonical wire-form transformation name shared across
// the cost meter, the audit envelope access_pattern field, and the
// transformation Prometheus label. DisplayName reproduces the Foundry
// doc verbatim so the Usage UI surfaces the same caption.
type CostEntry struct {
	Key                 string     `json:"key"`
	DisplayName         string     `json:"display_name"`
	Schema              SchemaKind `json:"schema"`
	ComputeSecondsPerGB uint32     `json:"compute_seconds_per_gb"`
}

// DownloadStream is the Foundry "All" row — applies to download /
// streaming of bytes regardless of media schema.
var DownloadStream = CostEntry{
	Key:                 "download_stream",
	DisplayName:         "Download / stream",
	Schema:              SchemaAll,
	ComputeSecondsPerGB: 2,
}

// CostTable mirrors the Rust COST_TABLE. Order matches the doc
// top-to-bottom so a screen-by-screen review against the upstream
// rendering catches a missing or reordered row instantly.
//
// DO NOT REORDER without updating the parity tests on both sides.
var CostTable = []CostEntry{
	// ── All ───────────────────────────────────────────────────────
	DownloadStream,
	// ── Images ────────────────────────────────────────────────────
	{Key: "generate_pdf", DisplayName: "Generate PDF", Schema: SchemaImage, ComputeSecondsPerGB: 40},
	{Key: "thumbnail", DisplayName: "Thumbnail (alias of Resize)", Schema: SchemaImage, ComputeSecondsPerGB: 40},
	{Key: "resize", DisplayName: "Resize", Schema: SchemaImage, ComputeSecondsPerGB: 40},
	{Key: "resize_within_bounding_box", DisplayName: "Resize within bounding box", Schema: SchemaImage, ComputeSecondsPerGB: 40},
	{Key: "rotate", DisplayName: "Rotate", Schema: SchemaImage, ComputeSecondsPerGB: 40},
	{Key: "adjust_contrast", DisplayName: "Adjust contrast", Schema: SchemaImage, ComputeSecondsPerGB: 75},
	{Key: "annotate", DisplayName: "Annotate", Schema: SchemaImage, ComputeSecondsPerGB: 75},
	{Key: "crop", DisplayName: "Crop / chip", Schema: SchemaImage, ComputeSecondsPerGB: 75},
	{Key: "encryption_decryption", DisplayName: "Encryption / decryption", Schema: SchemaImage, ComputeSecondsPerGB: 75},
	{Key: "geo_tile", DisplayName: "Geo tile", Schema: SchemaImage, ComputeSecondsPerGB: 75},
	{Key: "grayscale", DisplayName: "Grayscale", Schema: SchemaImage, ComputeSecondsPerGB: 75},
	{Key: "render_dicom_image_layer", DisplayName: "Render DICOM image layer", Schema: SchemaImage, ComputeSecondsPerGB: 75},
	{Key: "ocr", DisplayName: "Extract text (OCR)", Schema: SchemaImage, ComputeSecondsPerGB: 275},
	{Key: "embedding", DisplayName: "Generate embedding", Schema: SchemaImage, ComputeSecondsPerGB: 275},
	{Key: "ocr_v2", DisplayName: "Extract text v2 (OCR)", Schema: SchemaImage, ComputeSecondsPerGB: 275},
	{Key: "extract_layout_aware_v2", DisplayName: "Extract layout aware v2", Schema: SchemaImage, ComputeSecondsPerGB: 275},
	// ── Audio ─────────────────────────────────────────────────────
	{Key: "chunk", DisplayName: "Chunk", Schema: SchemaAudio, ComputeSecondsPerGB: 75},
	{Key: "channel_select", DisplayName: "Select channel", Schema: SchemaAudio, ComputeSecondsPerGB: 75},
	{Key: "hls_stream", DisplayName: "Stream with HLS", Schema: SchemaAudio, ComputeSecondsPerGB: 75},
	{Key: "transcode", DisplayName: "Transcode", Schema: SchemaAudio, ComputeSecondsPerGB: 75},
	{Key: "waveform", DisplayName: "Waveform generation", Schema: SchemaAudio, ComputeSecondsPerGB: 75},
	{Key: "transcription", DisplayName: "Transcription", Schema: SchemaAudio, ComputeSecondsPerGB: 275},
	// ── Video ─────────────────────────────────────────────────────
	{Key: "scene_frames_timestamps", DisplayName: "Get timestamps for scene frames", Schema: SchemaVideo, ComputeSecondsPerGB: 40},
	{Key: "extract_audio", DisplayName: "Extract audio", Schema: SchemaVideo, ComputeSecondsPerGB: 75},
	{Key: "extract_frames", DisplayName: "Extract frames at timestamp", Schema: SchemaVideo, ComputeSecondsPerGB: 75},
	{Key: "video_chunk", DisplayName: "Chunk", Schema: SchemaVideo, ComputeSecondsPerGB: 275},
	{Key: "scene_frames_all", DisplayName: "Extract all scene frames", Schema: SchemaVideo, ComputeSecondsPerGB: 275},
	{Key: "video_hls_stream", DisplayName: "Stream with HLS", Schema: SchemaVideo, ComputeSecondsPerGB: 275},
	{Key: "video_transcode", DisplayName: "Transcode", Schema: SchemaVideo, ComputeSecondsPerGB: 275},
	// ── Documents ─────────────────────────────────────────────────
	{Key: "pdf_page_dimensions", DisplayName: "Get PDF page dimensions", Schema: SchemaDocument, ComputeSecondsPerGB: 40},
	{Key: "render_page", DisplayName: "Render page as image", Schema: SchemaDocument, ComputeSecondsPerGB: 40},
	{Key: "render_page_bbox", DisplayName: "Render page as image within bounding box", Schema: SchemaDocument, ComputeSecondsPerGB: 40},
	{Key: "extract_text_raw", DisplayName: "Extract all text (raw)", Schema: SchemaDocument, ComputeSecondsPerGB: 75},
	{Key: "extract_form_fields", DisplayName: "Extract form fields", Schema: SchemaDocument, ComputeSecondsPerGB: 75},
	{Key: "extract_toc", DisplayName: "Extract table of contents", Schema: SchemaDocument, ComputeSecondsPerGB: 75},
	{Key: "extract_text_pages_raw", DisplayName: "Extract text from pages to array (raw)", Schema: SchemaDocument, ComputeSecondsPerGB: 75},
	{Key: "extract_text_on_page_raw", DisplayName: "Extract text on page (raw)", Schema: SchemaDocument, ComputeSecondsPerGB: 75},
	{Key: "slice_pdf_range", DisplayName: "Slice PDF range", Schema: SchemaDocument, ComputeSecondsPerGB: 75},
	{Key: "doc_ocr", DisplayName: "Extract text (OCR)", Schema: SchemaDocument, ComputeSecondsPerGB: 275},
	{Key: "extract_text_pages_ocr", DisplayName: "Extract text from pages to array (OCR)", Schema: SchemaDocument, ComputeSecondsPerGB: 275},
	{Key: "layout_aware_v2", DisplayName: "Extract layout aware text (raw / OCR)", Schema: SchemaDocument, ComputeSecondsPerGB: 275},
	{Key: "vlm_extract", DisplayName: "Extract text using VLM", Schema: SchemaDocument, ComputeSecondsPerGB: 275},
	// ── Spreadsheets ──────────────────────────────────────────────
	{Key: "render_sheet", DisplayName: "Convert sheet to JSON", Schema: SchemaSpreadsheet, ComputeSecondsPerGB: 275},
}

// RatePerGB returns the compute-seconds-per-GB rate for a transformation
// key, or (0, false) if the key is not in the published table — callers
// must refuse to charge for an off-table transformation rather than
// guess. The runtime catalog is the single source of truth for which
// keys are billable.
func RatePerGB(key string) (uint32, bool) {
	for i := range CostTable {
		if CostTable[i].Key == key {
			return CostTable[i].ComputeSecondsPerGB, true
		}
	}
	return 0, false
}

// ChargeComputeSeconds returns the compute-seconds charged for processing
// `bytes` through the transformation identified by `key`. Rounds **up**
// to the nearest whole compute-second so a 1-byte invocation still
// charges at least one compute-second. Returns (0, false) if `key` is
// not in the table.
func ChargeComputeSeconds(key string, bytes uint64) (uint64, bool) {
	rate, ok := RatePerGB(key)
	if !ok {
		return 0, false
	}
	const gb uint64 = 1024 * 1024 * 1024
	rate64 := uint64(rate)
	// Saturating multiply matches the Rust impl — clamp to MaxUint64 on
	// overflow so a multi-exabyte input still lands a finite charge.
	var prod uint64
	if rate64 != 0 && bytes > (^uint64(0))/rate64 {
		prod = ^uint64(0)
	} else {
		prod = rate64 * bytes
	}
	charged := (prod + gb - 1) / gb
	if bytes == 0 {
		return 0, true
	}
	if charged < 1 {
		charged = 1
	}
	return charged, true
}

// Entry returns the row for a key, including its display name and
// schema. Backs the /usage REST surface and the Usage UI tab.
func Entry(key string) (CostEntry, bool) {
	for i := range CostTable {
		if CostTable[i].Key == key {
			return CostTable[i], true
		}
	}
	return CostEntry{}, false
}

// EntriesFor returns every transformation belonging to a given schema
// plus the All rows. Backs the /usage JSON breakdown so the UI can
// populate the per-kind chart without knowing the table layout.
func EntriesFor(schema SchemaKind) []CostEntry {
	out := make([]CostEntry, 0, len(CostTable))
	for i := range CostTable {
		if CostTable[i].Schema == schema || CostTable[i].Schema == SchemaAll {
			out = append(out, CostTable[i])
		}
	}
	return out
}
